### PR TITLE
[UIE-128] Move append to core-utils

### PR DESCRIPTION
**Pull request by nawatts**
As part of splitting up the libs/utils module and packaging dependencies of the Environments page, this moves the `append` function to the core-utils package.

### CODE REVIEW

File: packages/core-utils/src/fp-utils.ts
Comment (nawatts) [inline packages/core-utils/src/fp-utils.ts:6]: It appeared that the type parameter didn't carry through `safeCurry`, so added a type with both signatures here.

File: src/analysis/AnalysisNotificationManager.ts
Comment (nawatts) [inline src/analysis/AnalysisNotificationManager.ts:4]: Went ahead and updated this import path since packaging Environments dependencies was a motivation.

File: src/import-data/ImportData.ts
Comment (nawatts) [inline src/import-data/ImportData.ts]: A type parameter is necessary here because `newJob` is an `AzureAsyncImportJob`, but we're appending to a more general list of `AsyncImportJob[]`.

File: src/libs/state.ts
Comment (nawatts) [inline src/libs/state.ts:262]: Need types here because TypeScript wouldn't allow appending a number or string to a list of `unknown`.

File: src/workspaces/ShareWorkspaceModal/ShareWorkspaceModal.ts
Comment (nawatts) [inline src/workspaces/ShareWorkspaceModal/ShareWorkspaceModal.ts:103]: Not entirely sure what to do here... Should the appended entry should have all the fields specified by AccessEntry? Should AccessEntry have optional fields? So I went with "no change" and left a comment for the Workspaces team.

File: packages/core-utils/src/fp-utils.ts
Comment (nawatts) [inline packages/core-utils/src/fp-utils.ts:1]: Re: adding an fp-utils module... I think deciding if or how to move away from an FP style is a different project and I didn't want it to hold up packaging.

File: src/workspaces/ShareWorkspaceModal/ShareWorkspaceModal.ts
Comment (cahrens) [reply src/workspaces/ShareWorkspaceModal/ShareWorkspaceModal.ts:103]: We are just going to show the default state for adding a user. I think `pending` should be marked as optional because only the server really knows that (we could mark a new user as "pending", but that will change the UI somewhat, and we already show new entries in green). And we should include `canCompute` and `canShare` with values of false since that is what we show as the state for a new user.

image

File: src/import-data/ImportData.ts
Comment (jladieu) [inline src/import-data/ImportData.ts]: Would you consider doing a direct import of `append` here like you did in index.ts to lower the friction of removing the `*` import?

**Review by jladieu — APPROVED** [review]

File: src/workspaces/ShareWorkspaceModal/ShareWorkspaceModal.ts
Comment (cahrens) [reply src/workspaces/ShareWorkspaceModal/ShareWorkspaceModal.ts:103]: But we can also do this ourselves…. I'm approving from a Workspaces CODEOWNER.

**Review by cahrens — APPROVED** [review]

File: src/import-data/ImportData.ts
Comment (nawatts) [reply src/import-data/ImportData.ts]: Yes. Long term, we want to move all imports that way. Generally, I've been doing these moves in two steps: the first for the move itself and any necessary changes and the second for cleaning up import paths in the Terra UI app, since that second step usually drags in more code owners and has a higher chance of merge conflicts. I broke that habit here for the analysis modules since removing their dependencies on libs/utils was the main motivation for doing this. Admittedly, I've gotten lazy about the import path cleanups, but it's on my todo list!

Comment (nawatts) [reply src/import-data/ImportData.ts]: But I guess there's no reason not to go ahead and update import paths in files I'm touching anyway. Done in https://github.com/DataBiosphere/terra-ui/pull/4647/commits/1d57c8ce3a29c7505c737bd0880d28a292a59ed9.

**Review by LizBaldo — APPROVED** [review]
Analysis file looks good 👍

File: packages/core-utils/src/fp-utils.ts
Comment (msilva-broad) [reply packages/core-utils/src/fp-utils.ts:6]: FYI, safeCurry wasn't used...

`export const append = _.curry((value, arr) => _.concat(arr, [value]));`

safeCurry should do the same function overload type flow you are establishing here...  unless you tried that and it didn't work?

Comment (nawatts) [reply packages/core-utils/src/fp-utils.ts:6]: Yeah, `safeCurry` didn't work with the type parameter.

```ts
export const append = safeCurry(<T>(value: T, arr: T[]) => _.concat(arr, [value]));
const appendFoo = append('foo');
// type of appendFoo is CurriedFunction1<unknown[], unknown[]>
```

Comment (msilva-broad) [reply packages/core-utils/src/fp-utils.ts:6]: what if you included the return type in the inner signature?  lodash functions like to any-sledgehammer types

```
export const append = safeCurry(<T>(value: T, arr: T[]): T[] => _.concat(arr, [value]));
const appendFoo = append('foo');
```

Comment (nawatts) [reply packages/core-utils/src/fp-utils.ts:6]: With that, `appendFoo`'s type is `any`.

Comment (nawatts) [reply packages/core-utils/src/fp-utils.ts:6]: Same result.

Comment (msilva-broad) [reply packages/core-utils/src/fp-utils.ts:6]: uggg.... its Friday and I'm being forced to think about how to better type-flow curried functions... ARG!!!!

Comment (nawatts) [reply packages/core-utils/src/fp-utils.ts:6]: Is there an issue with the `AppenFn` type added here? It seems fine to me if typing generic curried is a little verbose, especially if we want to limit curried functions.

**Comment by nawatts** [comment]
Closing this after talking to @msilva-broad. Holding off on promoting a curried function to core-utils. Replacements to follow...

**Pull request closed without merging**
